controllers/spifilecontentrequest_controller.go

```diff
@@ -0,0 +1,202 @@
+				}
+			}
+			return ret
+		})).
+		Complete(r)
+	if err != nil {
+		err = fmt.Errorf("failed to build the controller manager: %w", err)
+	}
+	return err
+
+}
+
+func (r *SPIFileContentRequestReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	ctx = infrastructure.InitKcpControllerContext(ctx, req)
+	lg := log.FromContext(ctx)
+
+	request := api.SPIFileContentRequest{}
+
+	if err := r.Get(ctx, req.NamespacedName, &request); err != nil {
+		if errors.IsNotFound(err) {
+			lg.Info("object not found")
+			return ctrl.Result{}, nil
+		}
+
+		lg.Error(err, "failed to get the object")
+		return ctrl.Result{}, fmt.Errorf("failed to read the object: %w", err)
+	}
+
+	if request.DeletionTimestamp != nil {
+		lg.Info("file request object is being deleted")
+		return ctrl.Result{}, nil
+	}
+
+	//TODO: lifetime cleanup
+
+	if request.Status.Phase == "" {
+		request.Status.Phase = api.SPIFileContentRequestPhaseAwaitingTokenData
+	}
+
+	if request.Status.LinkedBindingName == "" {
+		if err := r.createAndLinkBinding(ctx, &request); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create the object: %w", err)
+		}
+	} else {
+		binding := &api.SPIAccessTokenBinding{}
+		if err := r.Client.Get(ctx, client.ObjectKey{Name: request.Status.LinkedBindingName, Namespace: request.Namespace}, binding); err != nil {
+			if errors.IsNotFound(err) {
+				request.Status.LinkedBindingName = ""
+				r.updateFileRequestStatusError(ctx, &request, err)
+			}
+		}
+		//binding not yet fully ready to work with, let's try next time
+		if binding.Status.Phase == "" {
+			return reconcile.Result{}, nil
+		}
+
+		//binding not injected yet, let's just synchronize URL's and that's it
+		if binding.Status.Phase == api.SPIAccessTokenBindingPhaseAwaitingTokenData {
+			request.Status.OAuthUrl = binding.Status.OAuthUrl
+			request.Status.UploadUrl = binding.Status.UploadUrl
+		} else if binding.Status.Phase == api.SPIAccessTokenBindingPhaseInjected {
+			contents, err := gitfile.GetFileContents(ctx, r.Client, request.Namespace, binding.Status.SyncedObjectRef.Name, request.Spec.RepoUrl, request.Spec.FilePath, request.Spec.Ref)
+			if err != nil {
+				r.updateFileRequestStatusError(ctx, &request, err)
+				return reconcile.Result{}, err
+			}
+			fileBytes, err := io.ReadAll(contents)
+			if err != nil {
+				r.updateFileRequestStatusError(ctx, &request, err)
+				return reconcile.Result{}, err
+			}
+			request.Status.ContentEncoding = "base64"
+			request.Status.Content = base64.StdEncoding.EncodeToString(fileBytes)
+			request.Status.Phase = api.SPIFileContentRequestPhaseDelivered
+		} else {
+			err := fmt.Errorf("linked binding is in error state: %s", binding.Status.ErrorMessage)
+			r.updateFileRequestStatusError(ctx, &request, err)
+			return reconcile.Result{}, err
+		}
+	}
+	if err := r.Client.Status().Update(ctx, &request); err != nil {
+		log.FromContext(ctx).Error(err, "failed to update the file request status", "error", err)
```

Comment:
Failure to update the status should re-run the reconciliation, shouldn't it?

Reply:
added requeue